File: core/src/lib_impl/pyo3/container.rs
//! [`PyO3`](pyo3) implementation for [`Container`].
use crate::project::{Container, StandardProperties};
use crate::types::ResourceId;
use pyo3::exceptions::PyRuntimeError;
use pyo3::prelude::*;
use pythonize::pythonize;
use std::collections::{HashMap, HashSet};

// @todo: Unify implementations of property getters with Asset.
#[pymethods]
impl Container {
    // properties
    #[getter(_id)]
    fn py_id(&self) -> String {
        self.rid.to_string()
    }

    #[getter(_properties)]
    fn py_properties(&self) -> StandardProperties {
        self.properties.clone()
    }

    #[getter(name)]
    fn py_name(&self) -> Option<String> {
        self.properties.name.clone()
    }

    #[getter(type)]
    fn py_kind(&self) -> Option<String> {
        self.properties.kind.clone()
    }

    #[getter(tags)]
    fn py_tags(&self) -> HashSet<String> {
        self.properties.tags.clone().into_iter().collect()
    }

    // @todo: Provide inherited and native metadata separated.
    /// Returns all metadata.
    #[getter(metadata)]
    fn py_metadata(&self, py: Python<'_>) -> PyResult<HashMap<String, PyObject>> {
        let mut md = HashMap::with_capacity(self.properties.metadata.len());
        for (k, v) in self.properties.metadata.clone() {
            let val = pythonize(py, &v);
            if val.is_err() {
                return Err(PyRuntimeError::new_err(format!(
                    "Could not convert metadata of key `{}`: {:?}",
                    k, val
                )));
            }

            let val = val.unwrap();
            md.insert(k, val);
        }

        Ok(md)
    }

    // others
    #[getter(children)]
    fn py_children(&self) -> HashSet<ResourceId> {
        self.children.clone().into_keys().collect()
    }

    #[getter(assets)]
    fn py_assets(&self) -> HashSet<ResourceId> {
        self.assets.clone().into_keys().collect()
    }

    #[getter(parent)]
    fn py_parent(&self) -> Option<ResourceId> {
        self.parent.clone()
    }
}

#[cfg(test)]
#[path = "./container_test.rs"]
mod container_test;

File: lang/python/src/thot/container.py
from thot import _LEGACY_
from .types import OptStr, Tags, Metadata
from .asset import Asset

if _LEGACY_:
    from typing import List
    Assets = List[Asset]
else:
    Assets = list[Asset]

class Container:
    """
    A Container.
    """
    def __init__(
        self,
        rid: str,
        name: OptStr = None,
        type: OptStr = None,
        description: OptStr = None,
        tags: Tags = [],
        metadata: Metadata = {},
        assets: Assets = []
    ):
        """
        Create a new Container.
        """
        self._rid: str = rid
        self._name: OptStr = name
        self._type: OptStr = type
        self._description: OptStr = description
        self._tags: Tags = tags
        self._metadata: Metadata = metadata
        self._assets: Assets = assets
    
    @property
    def name(self) -> OptStr:
        """
        Returns:
            OptStr: Container's name.
        """
        return self._name
    
    @property
    def type(self) -> OptStr:
        """
        Returns:
            OptStr: Container's type.
        """
        return self._type
    
    @property
    def description(self) -> OptStr:
        """
        Returns:
            OptStr: Container's description.
        """
        return self._description
    
    @property
    def tags(self) -> Tags:
        """
        Returns:
            Tags: Container's tags.
        """
        return self._tags
    
    @property
    def metadata(self) -> Metadata:
        """
        Returns:
            Metadata: Container's metadata.
        """
        return self._metadata
        
    @property
    def assets(self) -> Assets:
        """
        Returns:
            list[Asset]: Container's Assets.
        """
        return self._assets